### PR TITLE
improvement: refactor Tendermint logger by using zerolog

### PR DESCRIPTION
**Pull request by alexanderbez**
Remove a lot of custom logging code in favor of a simple wrapper around [zerolog](https://github.com/rs/zerolog). The interface and auxiliary types (testing and nop) remain.

NOTE: This removes the ability for Tendermint's default logger to provide field filtering, e.g. `abci:info`.

### CODE REVIEW

File: libs/log/default.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+// Since zerolog supports typed structured logging and it is difficult to reflect
+// that in a generic interface, all logging methods accept a series of key/value
+// pair tuples, where the key must be a string.
+type DefaultLogger struct {
+	zerolog.Logger
+
+	trace bool
+}
+
+func NewDefaultLogger(format, level string, trace bool) (DefaultLogger, error) {
```
Comment (tychoish) [inline libs/log/default.go:28]: can't this just return an object of our Logger's interface type and we can avoid exporting `DefaultLogger` as a type?

File: libs/log/default.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+// DefaultLogger implements a default logger that can be used within Tendermint
+// and that fulfills the Logger interface. The underlying logging provider is a
+// zerolog logger that supports typical log levels along with JSON and plain/text
+// log formats.
+//
+// Since zerolog supports typed structured logging and it is difficult to reflect
+// that in a generic interface, all logging methods accept a series of key/value
+// pair tuples, where the key must be a string.
+type DefaultLogger struct {
```
Comment (tychoish) [inline libs/log/default.go:22]: I'd just make a constructor that takes a zerolog.Logger and returns one of our instances.

**Review by tychoish — CHANGES_REQUESTED** [review]

File: libs/log/default.go (hunk 1 above)
Comment (alexanderbez) [reply libs/log/default.go:28]: Yes, but best to return concrete types and not interfaces where possible.

File: libs/log/default.go (hunk 2 above)
Comment (alexanderbez) [reply libs/log/default.go:22]: We don't want folks having to even interact with or create zerolog. But we can have a constructor that accepts one for convenience.

File: libs/log/default.go (hunk 1 above)
Comment (alexanderbez) [reply libs/log/default.go:28]: That being said, I'm not hard-pressed on this. I mainly just wanted to remove all the existing logging cruft.

**Review by tychoish — APPROVED** [review]

File: libs/log/default.go (hunk 2 above)
Comment (tychoish) [reply libs/log/default.go]: yeah, that makes sense eventually. 

keeping this type public, in effect exposes a zerolog implementation, so I think it's good to make this private regardless.

**Comment by tychoish** [comment]
Is there some conversation somewhere that I've missed as to why this logger is better than other loggers, or where we decided to change?

**Comment by marbar3778** [comment]
> Is there some conversation somewhere that I've missed as to why this logger is better than other loggers, or where we decided to change?

https://github.com/tendermint/tendermint/issues/3428 although the title says zap the conversation says zero log. Also apps can easily override, and many do, the logger. I would say this is more of reducing maintenance burden of a custom logger.

**Comment by alexanderbez** [comment]
No major discussions @tychoish. We expose a `Logger` interface that applications can use to implement/provide whatever loggers they like.

Tendermint implements a "default" one, but it was archaic and very slow compared to zerolog, which is very performant and has good ergonomics.

File: libs/log/default.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+func (l defaultLogger) With(keyVals ...interface{}) Logger {
+	return defaultLogger{
+		Logger: l.Logger.With().Fields(getLogFields(keyVals...)).Logger(),
+		trace:  l.trace,
+	}
+}
+
+func getLogFields(keyVals ...interface{}) map[string]interface{} {
+	if len(keyVals)%2 != 0 {
+		return nil
```
Comment (tychoish) [inline libs/log/default.go:89]: doesn't this panic at the moment?

File: libs/log/default.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	fields := make(map[string]interface{})
+	for i := 0; i < len(keyVals); i += 2 {
+		fields[keyVals[i].(string)] = keyVals[i+1]
```
Comment (tychoish) [inline libs/log/default.go:94]: I worry about the type assertion. why not just `fmt.Sprint()`

File: libs/log/default.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		trace:  l.trace,
+	}
+}
+
+func getLogFields(keyVals ...interface{}) map[string]interface{} {
+	if len(keyVals)%2 != 0 {
+		return nil
+	}
+
+	fields := make(map[string]interface{})
```
Comment (tychoish) [inline libs/log/default.go:92]: this is likely going to be the bottleneck moving forward (converting the list to a map)

if nothing else we should prealloc the map based on the num of keys we know we'll have.

**Review by tychoish — COMMENTED** [review]
not that logging performance is really ever an issue, but...

File: libs/log/default.go (hunk 5 above)
Comment (alexanderbez) [reply libs/log/default.go:92]: I think that's fine, but yeah we can prealloc.

File: libs/log/default.go (hunk 3 above)
Comment (alexanderbez) [reply libs/log/default.go:89]: I'm not sure, but we should not.

File: libs/log/default.go (hunk 4 above)
Comment (alexanderbez) [reply libs/log/default.go:94]: We can use that.

File: libs/log/default.go [hunk 6]
```diff
@@ -89,9 +89,9 @@ func getLogFields(keyVals ...interface{}) map[string]interface{} {
 		return nil
 	}
 
-	fields := make(map[string]interface{})
+	fields := make(map[string]interface{}, len(keyVals))
 	for i := 0; i < len(keyVals); i += 2 {
-		fields[keyVals[i].(string)] = keyVals[i+1]
+		fields[fmt.Sprintf("%v", keyVals[i])] = keyVals[i+1]
```
Comment (tychoish) [inline libs/log/default.go:94]: I think `fmt.Sprint()` rather than `fmt.Sprintf()` is the right call here, but, yeah.

**Comment by alexanderbez** [comment]
We need the logger to be thread-safe, which zerolog supports...I'm gonna put this PR on pause.

**Comment by marbar3778** [comment]
> not that logging performance is really ever an issue, but...

It has caused issues in the past at least in some testing areas.

Not sure that there needs to be a further discussion here, does there? The pr is already here, lets just wrap it up and merge it. Less code is always better, IMO.

**Comment by tychoish** [comment]
> It has caused issues in the past at least in some testing areas.

The testing logger problem is almost certainly not about the raw speed of the logger and rather about the volume of log messages when the system runs quickly (combined with the way that the testing logger writes to standard output, and the logger is multiplexed between tests. 

I agree, however, that this is a fine change.

**Comment by alexanderbez** [comment]
It's a win for two reasons: (1) We have a more performant logger (which does matter, albeit can be little), and (2) we removed +1k lines of useless code.

**Comment by tychoish** [comment]
Yeah, let's do it.

File: libs/log/logger.go [hunk 7]
```diff
@@ -2,29 +2,57 @@ package log
...
+	// logging facilities that support complex log parsing and querying.
+	LogFormatJSON string = "json"
 
-	kitlog "github.com/go-kit/kit/log"
+	// Supported loging levels
+	LogLevelDebug = "debug"
+	LogLevelInfo  = "info"
+	LogLevelWarn  = "warn"
+	LogLevelError = "error"
+	LogLevelFatal = "fatal"
```
Comment (marbar3778) [inline libs/log/logger.go:29]: Do we support warn and fatal?

**Review by marbar3778 — APPROVED** [review]
LGTM

File: go.mod [hunk 8]
```diff
@@ -31,6 +31,7 @@ require (
 	github.com/prometheus/client_golang v1.10.0
 	github.com/rcrowley/go-metrics v0.0.0-20200313005456-10cdbea86bc0
 	github.com/rs/cors v1.7.0
+	github.com/rs/zerolog v1.22.0
```
[the remark this answers is not in the document]
Comment (alexanderbez) [reply go.mod:34]: Yes, for metrics :-/

File: libs/log/logger.go (hunk 7 above)
Comment (alexanderbez) [reply libs/log/logger.go:29]: We don't technically, although I think we should support warn at the least. What do you think? I'll remove fatal.

Comment (marbar3778) [reply libs/log/logger.go]: Makes sense, do you need to update the interface to expose it as well?

**Comment by marbar3778** [comment]
> NOTE: This removes the ability for Tendermint's default logger to provide field filtering, e.g. `abci:info`.

docs need to be updated to reflect this. Mainly logging.md

**Pull request merged**